omas/omas_physics.py

```diff
@@ -82,6 +129,82 @@ def core_profiles_pressures(ods, update=True):
     return ods_p
 
 
+@add_to__ODS__
+def core_profiles_densities(ods, update=True):
+    '''
+    calculates density from density_thermal and density_fast
+
+    :param ods: input ods
+
+    :param update: operate in place
+
+    :return: updated ods
+    '''
+
+    ods_n = ods
+    if not update:
+        from omas import ODS
+        ods_n = ODS().copy_attrs_from(ods)
+
+    for time_index in ods['core_profiles']['profiles_1d']:
+        prof1d = ods['core_profiles']['profiles_1d'][time_index]
+        prof1d_n = ods_n['core_profiles']['profiles_1d'][time_index]
+
+        if not update:
+            prof1d_n['grid']['rho_tor_norm'] = prof1d['grid']['rho_tor_norm']
+
+        __zeros__ = 0.*prof1d['grid']['rho_tor_norm']
+
+        # electrons
+        prof1d_n['electrons']['density'] = copy.deepcopy(__zeros__)
+        for density in ['density_thermal','density_fast']:
+            if density in prof1d['electrons']:
+                 prof1d_n['electrons']['density'] += prof1d['electrons'][density]
+
+        # ions
+        for k in range(len(prof1d['ion'])):
+            prof1d_n['ion'][k]['density'] = copy.deepcopy(__zeros__)
+            for density in ['density_thermal','density_fast']:
+                if density in prof1d['ion'][k]:
+                    prof1d_n['ion'][k]['density'] += prof1d['ion'][k][density]
+    return ods_n
+
+@add_to__ODS__
+def core_profiles_zeff(ods, update=True, use_electrons_density=False):
+    '''
+    calculates effective charge
+
+    :param ods: input ods
+
+    :param update: operate in place
+
+    :param use_electrons_density: 
+            denominator core_profiles.profiles_1d.:.electrons.density 
+            instead of sum Z*n_i
+
+    :return: updated ods
+    '''
+
+    ods_z = core_profiles_densities(ods,update=update)
+
+    for time_index in ods['core_profiles']['profiles_1d']:
+        prof1d = ods['core_profiles']['profiles_1d'][time_index]
+        prof1d_z = ods_z['core_profiles']['profiles_1d'][time_index]
+
+        Z2n = 0.*prof1d_z['grid']['rho_tor_norm']
+        Zn  = 0.*prof1d_z['grid']['rho_tor_norm']
+
+        for k in range(len(prof1d['ion'])):
+            Z = prof1d['ion'][k]['element'][0]['z_n'] # from old ODS
```

Comment:
@bclyons12 any reason why you cannot simply work with `ods_z`? The atomic element `z_n` should be there.

Reply:
Nope, I could do that. I was just following the "read from old, write to new" logic of `core_profiles_pressures` and `core_profiles_densities`